main.c

```diff
@@ -177,26 +178,28 @@ void writeBuffer(struct vBuffer **buffer, AVFrame *frame){
 	(*buffer)->next->next = NULL;
 	vStream.bufferLength++;
 
-	vStream.semaphore = 0;
+	pthread_mutex_unlock(&vStream.mutex);
 }
 
 // Return frame buffer at the head of our queue
 struct vBuffer* readBuffer(struct vBuffer **buffer){
 	struct vBuffer *nextBuffer;
 
-	// Wait for the semaphore so we don't cause any issues
-	while (vStream.semaphore != 0 || (vStream.bufferLength <= 2 && !vStream.bufferEnd)) nanosleep(&wait, (struct timespec*) NULL);;
-	vStream.semaphore = 1;
-
-	// Update buffer and queue
-	nextBuffer = (*buffer)->next;
-	free((*buffer)->frame);
-	free(*buffer);
-	vStream.bufferLength--;
-	vStream.semaphore = 0;
+	// don't bother trying to do anything if there's no need to do work
+	while ((vStream.bufferLength <= 2 && !vStream.bufferEnd)) nanosleep(&wait, (struct timespec*) NULL);;
+	// Wait for the mutex so we don't cause any issues
+	while (pthread_mutex_trylock(&vStream.mutex) != 0) nanosleep(&wait, (struct timespec*) NULL);
+
+	if ((*buffer)->next != NULL) {
```

Comment:
I'm actually not sure we need this block... This might make us not read or free the last frame